File: DSA/VedantChouhan_590024563/Day-78/q2.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_BITS 31
#define MAX_NODES 3100000

typedef struct {
    int child[2];
} TrieNode;

TrieNode trie[MAX_NODES];
int nodeCount = 1;

/* Insert a number into the binary Trie */
void insert(int num) {
    int node = 0;

    for (int bit = MAX_BITS - 1; bit >= 0; bit--) {
        int b = (num >> bit) & 1;

        if (trie[node].child[b] == 0) {
            trie[node].child[b] = nodeCount++;
        }

        node = trie[node].child[b];
    }
}

/* Find the maximum XOR possible with num */
int getMaxXOR(int num) {
    int node = 0;
    int result = 0;

    for (int bit = MAX_BITS - 1; bit >= 0; bit--) {
        int b = (num >> bit) & 1;
        int opposite = 1 - b;

        if (trie[node].child[opposite] != 0) {
            result |= (1 << bit);
            node = trie[node].child[opposite];
        } else {
            node = trie[node].child[b];
        }
    }

    return result;
}

int main() {
    int n;

    printf("Enter number of elements: ");
    scanf("%d", &n);

    int nums[n];

    printf("Enter elements: ");

    for (int i = 0; i < n; i++) {
        scanf("%d", &nums[i]);
    }

    if (n < 2) {
        printf("Maximum XOR Pair: 0\n");
        return 0;
    }

    int answer = 0;

    /*
       Insert the first number.
       For every next number, find the best XOR
       with numbers already inserted.
    */
    insert(nums[0]);

    for (int i = 1; i < n; i++) {

        int current = getMaxXOR(nums[i]);

        if (current > answer)
            answer = current;

        insert(nums[i]);
    }

    printf("Maximum XOR Pair: %d\n", answer);

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day78/Q2.java
import java.util.*;

class Solution {

    static class TrieNode {
        TrieNode[] child = new TrieNode[2];
    }

    static TrieNode root = new TrieNode();

    static void insert(int num) {
        TrieNode curr = root;

        for (int i = 30; i >= 0; i--) {
            int bit = (num >> i) & 1;

            if (curr.child[bit] == null) {
                curr.child[bit] = new TrieNode();
            }

            curr = curr.child[bit];
        }
    }

    static int getMaxXOR(int num) {
        TrieNode curr = root;
        int xor = 0;

        for (int i = 30; i >= 0; i--) {
            int bit = (num >> i) & 1;
            int opposite = 1 - bit;

            if (curr.child[opposite] != null) {
                xor |= (1 << i);
                curr = curr.child[opposite];
            } else {
                curr = curr.child[bit];
            }
        }

        return xor;
    }

    public static void main(String[] args) {

        int[] arr = {1, 2, 3, 4};

        for (int num : arr) {
            insert(num);
        }

        int answer = 0;

        for (int num : arr) {
            answer = Math.max(answer, getMaxXOR(num));
        }

        System.out.println(answer);
    }
}